### ISSUE TITLE
Problem: query unparsing for Postgres 16 may be imperfect

### ISSUE THREAD

**Issue by yrashk** [omnigres/omnigres#411]
The original effort in e35b689680b7dfbb71e5da41b51bf03e1ae8563e added [deparse_16.c](https://github.com/omnigres/omnigres/blob/master/extensions/omni_sql/deparse_16.c), but it hasn't been maintained since and may not be perfect. It was made to be just sort of good enough to support Postgres 16.

Now, the original library (libpg_query) already has a `16-latest` branch and its own version of that [deparse file](https://github.com/pganalyze/libpg_query/blob/16-latest/src/postgres_deparse.c)

Proposed solution: review changes and update our deparse_16.c with the one in the project.

If, by any luck, some of the changes we did are beneficial to the project, we should contribute them back.

---

**Bounty size: $100**

[Read more about bounties](https://github.com/omnigres/omnigres/wiki/Bounties)

**Comment by Khalid6468** [comment]
Hi @yrashk I'm interested in working on this issue. Can you assign this to me?

**Comment by yrashk** [comment]
@Khalid6468 here you go!

**Comment by yrashk** [comment]
I will post the diff between our version of the deparsing for Postgres 15 and 16 here for ease of reference to see what was changed on our end:

```diff
--- extensions/omni_sql/deparse_15.c	2023-12-20 15:51:27
+++ extensions/omni_sql/deparse_16.c	2023-12-20 15:51:27
@@ -36,13 +36,14 @@
 #include <postgres.h>
 #include <fmgr.h>
 
-#if PG_MAJORVERSION_NUM == 15
+#if PG_MAJORVERSION_NUM == 16
 
 #include "catalog/index.h"
 #include "catalog/pg_am.h"
 #include "catalog/pg_attribute.h"
 #include "catalog/pg_class.h"
 #include "catalog/pg_trigger.h"
+#include "commands/defrem.h"
 #include "commands/trigger.h"
 #include "common/keywords.h"
 #include "common/kwlookup.h"
\ No newline at end of file
@@ -2885,6 +2886,7 @@
     break;
   case JOIN_SEMI:
   case JOIN_ANTI:
+  case JOIN_RIGHT_ANTI:
   case JOIN_UNIQUE_OUTER:
   case JOIN_UNIQUE_INNER:
     // Only used by the planner/executor, not seen in parser output
\ No newline at end of file
@@ -4435,7 +4437,17 @@
   ListCell *lc;
 
   appendStringInfoString(str, "PARTITION BY ");
-  appendStringInfoString(str, partition_spec->strategy);
+  switch (partition_spec->strategy) {
+    case PARTITION_STRATEGY_LIST:
+      appendStringInfoString(str, "list");
+      break;
+   case PARTITION_STRATEGY_RANGE:
+      appendStringInfoString(str, "range");
+      break;
+   case PARTITION_STRATEGY_HASH:
+      appendStringInfoString(str, "hash");
+      break;
+  }
 
   appendStringInfoChar(str, '(');
   foreach (lc, partition_spec->partParams) {
\ No newline at end of file
@@ -5434,9 +5446,6 @@
     else
       appendStringInfoString(str, "ADD COLUMN ");
     break;
-  case AT_AddColumnRecurse: /* internal to commands/tablecmds.c */
-    Assert(false);
-    break;
   case AT_AddColumnToView: /* implicitly via CREATE OR REPLACE VIEW */
     // Not present in raw parser output
     Assert(false);
\ No newline at end of file
@@ -5495,9 +5504,6 @@
     else
       appendStringInfoString(str, "DROP ");
     break;
-  case AT_DropColumnRecurse: /* internal to commands/tablecmds.c */
-    Assert(false);
-    break;
   case AT_AddIndex: /* add index */
     appendStringInfoString(str, "ADD INDEX ");
     break;
\ No newline at end of file
@@ -5507,9 +5513,6 @@
   case AT_AddConstraint: /* add constraint */
     appendStringInfoString(str, "ADD ");
     break;
-  case AT_AddConstraintRecurse: /* internal to commands/tablecmds.c */
-    Assert(false);
-    break;
   case AT_ReAddConstraint: /* internal to commands/tablecmds.c */
     Assert(false);
     break;
\ No newline at end of file
@@ -5524,9 +5527,6 @@
   case AT_ValidateConstraint: /* validate constraint */
     appendStringInfoString(str, "VALIDATE CONSTRAINT ");
     break;
-  case AT_ValidateConstraintRecurse: /* internal to commands/tablecmds.c */
-    Assert(false);
-    break;
   case AT_AddIndexConstraint: /* add constraint using existing index */
     // Not present in raw parser output
     Assert(false);
\ No newline at end of file
@@ -5534,9 +5534,6 @@
   case AT_DropConstraint: /* drop constraint */
     appendStringInfoString(str, "DROP CONSTRAINT ");
     break;
-  case AT_DropConstraintRecurse: /* internal to commands/tablecmds.c */
-    Assert(false);
-    break;
   case AT_ReAddComment:    /* internal to commands/tablecmds.c */
   case AT_ReAddStatistics: /* internal to commands/tablecmds.c */
     Assert(false);
\ No newline at end of file
@@ -7059,13 +7056,24 @@
 static void deparseGrantRoleStmt(StringInfo str, GrantRoleStmt *grant_role_stmt) {
   ListCell *lc;
 
+  DefElem *opt = (DefElem *)linitial(grant_role_stmt->opt);
+
   if (grant_role_stmt->is_grant)
     appendStringInfoString(str, "GRANT ");
   else
     appendStringInfoString(str, "REVOKE ");
 
-  if (!grant_role_stmt->is_grant && grant_role_stmt->admin_opt)
-    appendStringInfoString(str, "ADMIN OPTION FOR ");
+  if (!grant_role_stmt->is_grant) {
+    ListCell *item;
+    foreach(item, grant_role_stmt->opt) {
+      DefElem *opt = (DefElem *) lfirst(item);
+      char *optval = defGetString(opt);
+      if (strcmp(opt->defname, "admin") == 0) {
+        appendStringInfoString(str, "ADMIN OPTION FOR ");
+      }
+      // TODO: other options
+    }
+  }
 
   foreach (lc, grant_role_stmt->granted_roles) {
     deparseAccessPriv(str, castNode(AccessPriv, lfirst(lc)));
\ No newline at end of file
@@ -7082,8 +7090,17 @@
   deparseRoleList(str, grant_role_stmt->grantee_roles);
   appendStringInfoChar(str, ' ');
 
-  if (grant_role_stmt->is_grant && grant_role_stmt->admin_opt)
-    appendStringInfoString(str, "WITH ADMIN OPTION ");
+  if (!grant_role_stmt->is_grant) {
+    ListCell *item;
+    foreach(item, grant_role_stmt->opt) {
+      DefElem *opt = (DefElem *) lfirst(item);
+      char *optval = defGetString(opt);
+      if (strcmp(opt->defname, "admin") == 0) {
+        appendStringInfoString(str, "ADMIN OPTION FOR ");
+      }
+      // TODO: other options
+    }
+  }
 
   if (grant_role_stmt->grantor) {
     appendStringInfoString(str, "GRANTED BY ");
\ No newline at end of file
@@ -9528,5 +9545,5 @@
   return deparseRawStmt(str, raw_stmt);
 }
 
-#endif // PG_MAJORVERSION_NUM == 15
+#endif // PG_MAJORVERSION_NUM == 16
 // clang-format on
\ No newline at end of file
```

**Comment by Khalid6468** [comment]
> I will post the diff between our version of the deparsing for Postgres 15 and 16 here for ease of reference to see what was changed on our end:
> 
> ```diff
> --- extensions/omni_sql/deparse_15.c	2023-12-20 15:51:27
> +++ extensions/omni_sql/deparse_16.c	2023-12-20 15:51:27
> @@ -36,13 +36,14 @@
>  #include <postgres.h>
>  #include <fmgr.h>
>  
> -#if PG_MAJORVERSION_NUM == 15
> +#if PG_MAJORVERSION_NUM == 16
>  
>  #include "catalog/index.h"
>  #include "catalog/pg_am.h"
>  #include "catalog/pg_attribute.h"
>  #include "catalog/pg_class.h"
>  #include "catalog/pg_trigger.h"
> +#include "commands/defrem.h"
>  #include "commands/trigger.h"
>  #include "common/keywords.h"
>  #include "common/kwlookup.h"
> \ No newline at end of file
> @@ -2885,6 +2886,7 @@
>      break;
>    case JOIN_SEMI:
>    case JOIN_ANTI:
> +  case JOIN_RIGHT_ANTI:
>    case JOIN_UNIQUE_OUTER:
>    case JOIN_UNIQUE_INNER:
>      // Only used by the planner/executor, not seen in parser output
> \ No newline at end of file
> @@ -4435,7 +4437,17 @@
>    ListCell *lc;
>  
>    appendStringInfoString(str, "PARTITION BY ");
> -  appendStringInfoString(str, partition_spec->strategy);
> +  switch (partition_spec->strategy) {
> +    case PARTITION_STRATEGY_LIST:
> +      appendStringInfoString(str, "list");
> +      break;
> +   case PARTITION_STRATEGY_RANGE:
> +      appendStringInfoString(str, "range");
> +      break;
> +   case PARTITION_STRATEGY_HASH:
> +      appendStringInfoString(str, "hash");
> +      break;
> +  }
>  
>    appendStringInfoChar(str, '(');
>    foreach (lc, partition_spec->partParams) {
> \ No newline at end of file
> @@ -5434,9 +5446,6 @@
>      else
>        appendStringInfoString(str, "ADD COLUMN ");
>      break;
> -  case AT_AddColumnRecurse: /* internal to commands/tablecmds.c */
> -    Assert(false);
> -    break;
>    case AT_AddColumnToView: /* implicitly via CREATE OR REPLACE VIEW */
>      // Not present in raw parser output
>      Assert(false);
> \ No newline at end of file
> @@ -5495,9 +5504,6 @@
>      else
>        appendStringInfoString(str, "DROP ");
>      break;
> -  case AT_DropColumnRecurse: /* internal to commands/tablecmds.c */
> -    Assert(false);
> -    break;
>    case AT_AddIndex: /* add index */
>      appendStringInfoString(str, "ADD INDEX ");
>      break;
> \ No newline at end of file
> @@ -5507,9 +5513,6 @@
>    case AT_AddConstraint: /* add constraint */
>      appendStringInfoString(str, "ADD ");
>      break;
> -  case AT_AddConstraintRecurse: /* internal to commands/tablecmds.c */
> -    Assert(false);
> -    break;
>    case AT_ReAddConstraint: /* internal to commands/tablecmds.c */
>      Assert(false);
>      break;
> \ No newline at end of file
> @@ -5524,9 +5527,6 @@
>    case AT_ValidateConstraint: /* validate constraint */
>      appendStringInfoString(str, "VALIDATE CONSTRAINT ");
>      break;
> -  case AT_ValidateConstraintRecurse: /* internal to commands/tablecmds.c */
> -    Assert(false);
> -    break;
>    case AT_AddIndexConstraint: /* add constraint using existing index */
>      // Not present in raw parser output
>      Assert(false);
> \ No newline at end of file
> @@ -5534,9 +5534,6 @@
>    case AT_DropConstraint: /* drop constraint */
>      appendStringInfoString(str, "DROP CONSTRAINT ");
>      break;
> -  case AT_DropConstraintRecurse: /* internal to commands/tablecmds.c */
> -    Assert(false);
> -    break;
>    case AT_ReAddComment:    /* internal to commands/tablecmds.c */
>    case AT_ReAddStatistics: /* internal to commands/tablecmds.c */
>      Assert(false);
> \ No newline at end of file
> @@ -7059,13 +7056,24 @@
>  static void deparseGrantRoleStmt(StringInfo str, GrantRoleStmt *grant_role_stmt) {
>    ListCell *lc;
>  
> +  DefElem *opt = (DefElem *)linitial(grant_role_stmt->opt);
> +
>    if (grant_role_stmt->is_grant)
>      appendStringInfoString(str, "GRANT ");
>    else
>      appendStringInfoString(str, "REVOKE ");
>  
> -  if (!grant_role_stmt->is_grant && grant_role_stmt->admin_opt)
> -    appendStringInfoString(str, "ADMIN OPTION FOR ");
> +  if (!grant_role_stmt->is_grant) {
> +    ListCell *item;
> +    foreach(item, grant_role_stmt->opt) {
> +      DefElem *opt = (DefElem *) lfirst(item);
> +      char *optval = defGetString(opt);
> +      if (strcmp(opt->defname, "admin") == 0) {
> +        appendStringInfoString(str, "ADMIN OPTION FOR ");
> +      }
> +      // TODO: other options
> +    }
> +  }
>  
>    foreach (lc, grant_role_stmt->granted_roles) {
>      deparseAccessPriv(str, castNode(AccessPriv, lfirst(lc)));
> \ No newline at end of file
> @@ -7082,8 +7090,17 @@
>    deparseRoleList(str, grant_role_stmt->grantee_roles);
>    appendStringInfoChar(str, ' ');
>  
> -  if (grant_role_stmt->is_grant && grant_role_stmt->admin_opt)
> -    appendStringInfoString(str, "WITH ADMIN OPTION ");
> +  if (!grant_role_stmt->is_grant) {
> +    ListCell *item;
> +    foreach(item, grant_role_stmt->opt) {
> +      DefElem *opt = (DefElem *) lfirst(item);
> +      char *optval = defGetString(opt);
> +      if (strcmp(opt->defname, "admin") == 0) {
> +        appendStringInfoString(str, "ADMIN OPTION FOR ");
> +      }
> +      // TODO: other options
> +    }
> +  }
>  
>    if (grant_role_stmt->grantor) {
>      appendStringInfoString(str, "GRANTED BY ");
> \ No newline at end of file
> @@ -9528,5 +9545,5 @@
>    return deparseRawStmt(str, raw_stmt);
>  }
>  
> -#endif // PG_MAJORVERSION_NUM == 15
> +#endif // PG_MAJORVERSION_NUM == 16
>  // clang-format on
> \ No newline at end of file
> ```

Okay

**Comment by yrashk** [comment]
(Not saying the above changes are guaranteed to be accurate, but at least it shows what was the difference)

**Comment by yrashk** [comment]
Solved in #412. Congrats! 🎊
